dali/kernels/imgproc/jpeg/jpeg_distortion_gpu.cuh

```diff
@@ -0,0 +1,267 @@
+  };
+
+  const Surface2D<uint8_t> out = {
+    sample.out, sample.size, 3, sample.strides, 1
+  };
+
+  for (int pos_y = y_start; pos_y < block.end.y; pos_y += blockDim.y) {
+    for (int pos_x = x_start; pos_x < block.end.x; pos_x += blockDim.x) {
+      int y = pos_y << vert_subsample;
+      int x = pos_x << horz_subsample;
+      auto ycbcr = rgb_to_ycbcr_subsampled<horz_subsample, vert_subsample, T>(ivec2{x, y}, in);
+      ycbcr_to_rgb_subsampled<horz_subsample, vert_subsample, T>(ivec2{x, y}, out, ycbcr);
+    }
+  }
+}
+
+
+
+template <bool horz_subsample, bool vert_subsample>
+__global__ void JpegCompressionDistortion(const SampleDesc *samples,
+                                          const kernels::BlockDesc<2> *blocks) {
+  using T = uint8_t;
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  int y_start = threadIdx.y + block.start.y;
+  int x_start = threadIdx.x + block.start.x;
+  if (y_start >= block.end.y || x_start >= block.end.x) {
+    return;
+  }
+
+  // Assuming CUDA block is 32x16, leading to a 2x4 grid of chroma blocks
+  // and up to 4x8 blocks of luma.
+  __shared__ T luma_blk[2 << vert_subsample][4 << horz_subsample][8][9];
+  __shared__ T cb_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
+  __shared__ T cr_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
+
+  int chroma_x = threadIdx.x & 7;  // % 8
+  int chroma_y = threadIdx.y & 7;  // % 8
+  ivec2 chroma_blk_idx{threadIdx.x >> 3, threadIdx.y >> 3};  // / 8
+
+  int luma_x = threadIdx.x << horz_subsample;
+  int luma_y = threadIdx.y << vert_subsample;
+  ivec2 luma_blk_idx{luma_x >> 3, luma_y >> 3};  // / 8
+  luma_x = luma_x & 7;  // % 8
+  luma_y = luma_y & 7;  // % 8
+
+  const Surface2D<const uint8_t> in = {
+    sample.in, sample.size, 3, sample.strides, 1
+  };
+
+  const Surface2D<uint8_t> out = {
+    sample.out, sample.size, 3, sample.strides, 1
+  };
+
+  T (&luma)[8][9] = luma_blk[luma_blk_idx.y][luma_blk_idx.x];
+  T (&cb)[8][9] = cb_blk[chroma_blk_idx.y][chroma_blk_idx.x];
+  T (&cr)[8][9] = cr_blk[chroma_blk_idx.y][chroma_blk_idx.x];
+
+  for (int pos_y = y_start; pos_y < block.end.y; pos_y += blockDim.y) {
+    for (int pos_x = x_start; pos_x < block.end.x; pos_x += blockDim.x) {
+      int y = pos_y << vert_subsample;
+      int x = pos_x << horz_subsample;
+      ivec2 offset{x, y};
+
+      auto ycbcr = rgb_to_ycbcr_subsampled<horz_subsample, vert_subsample, T>(offset, in);
+      luma[luma_y][luma_x] = ycbcr.luma[0];
+      cb[chroma_y][chroma_x] = ycbcr.cb;
+      cr[chroma_y][chroma_x] = ycbcr.cr;
+      if (horz_subsample && vert_subsample) {
+        luma[luma_y][luma_x + 1] = ycbcr.luma[1];
+        luma[luma_y + 1][luma_x] = ycbcr.luma[2];
+        luma[luma_y + 1][luma_x + 1] = ycbcr.luma[3];
+      } else if (horz_subsample) {
+        luma[luma_y][luma_x + 1] = ycbcr.luma[1];
+      } else if (vert_subsample) {
+        luma[luma_y + 1][luma_x] = ycbcr.luma[1];
+      }
```

Comment:
This would be shorter - and hopefully the compiler would be smart enough to unroll it.
```suggestion
      for (int i = 0, k = 0; i < vert_subsample+1; i++)
        for (int j = 0; j < horz_subsample+1; j++, k++)
          luma[luma_y + i][luma_x + j] = ycbcr.luma[k];
```